### ISSUE TITLE
DPO rewards stucks at zero 

### ISSUE THREAD

**Issue by pankayaraj** [huggingface/trl#1311]
While finetuning Llama from an SFT model trained with lora config I get this type of behavior where both the rewards stay at 0 and the loss never goes down

15 {'loss': 0.6932, 'learning_rate': 1.3724429223744293e-05, 'rewards/chosen': 0.0, 'rewards/rejected': 0.0, 'rewards/accuracies': 0.0, 'rewards/margins': 0.0, 'logps/rejected': -93.45967864990234, 'logps/chosen': -76.77323150634766, 'logits/rejected': -1.6269360780715942, 'logits/chosen': -1.6115689277648926, 'epoch': 0.38}
16 {'loss': 0.6932, 'learning_rate': 1.3670776255707763e-05, 'rewards/chosen': 0.0, 'rewards/rejected': 0.0, 'rewards/accuracies': 0.0, 'rewards/margins': 0.0, 'logps/rejected': -96.17804718017578, 'logps/chosen': -74.35616302490234, 'logits/rejected': -1.6258790493011475, 'logits/chosen': -1.5984680652618408, 'epoch': 0.41}
17 {'loss': 0.6932, 'learning_rate': 1.3617123287671234e-05, 'rewards/chosen': 0.0, 'rewards/rejected': 0.0, 'rewards/accuracies': 0.0, 'rewards/margins': 0.0, 'logps/rejected': -98.59229278564453, 'logps/chosen': -73.85718536376953, 'logits/rejected': -1.6299934387207031, 'logits/chosen': -1.606400489807129, 'epoch': 0.45}
18 {'loss': 0.6932, 'learning_rate': 1.3563470319634702e-05, 'rewards/chosen': 0.0, 'rewards/rejected': 0.0, 'rewards/accuracies': 0.0, 'rewards/margins': 0.0, 'logps/rejected': -93.72036743164062, 'logps/chosen': -79.21975708007812, 'logits/rejected': -1.6140862703323364, 'logits/chosen': -1.5989283323287964, 'epoch': 0.49}

I used the following training arguments.

I tried with both fp16 and bf16

training_args = TrainingArguments(
        per_device_train_batch_size=1,
        gradient_accumulation_steps=32,
        remove_unused_columns=False,
        num_train_epochs=epochs, 
        output_dir=save_dir,
        save_steps=1500,
        logging_first_step=True,
        logging_steps=5, 
        learning_rate=1.41e-5,
        optim="rmsprop",
        warmup_steps=0,
        #bf16=True,
        #fp16=True,
        )

**Comment by younesbelkada** [comment]
Hmm, that's a bit weird, cc @kashif if you have any idea
not sure if this is a duplicate of https://github.com/huggingface/trl/issues/1236 - might be a HP issue? (that issue though points out about the loss, not the reward)

**Comment by AlexiaJM** [comment]
@pankayaraj Did you find a solution? I have the same bug. 

Mine is at loss=0.6931 instead of .6932 and the logps/rejected goes down to crazy numbers near the end ('logps/rejected': -36880.9296875).

**Comment by AlexiaJM** [comment]
I solved the issue on my side, by installing the dev-build from github of 'trl' and the latest pip version of 'datasets'.

**Comment by kashif** [comment]
Great! Let me know how it goes!On 14. Feb 2024, at 22:31, Alexia Jolicoeur-Martineau ***@***.***> wrote:﻿
I solved the issue on my side, by installing the dev-build from github of 'trl' and the latest pip version of 'datasets'.

**Comment by 0nutation** [comment]
@pankayaraj  Did you find a solution? I have the same bug. My dpo training loss stucks at 0.6931.

**Comment by 0nutation** [comment]
@AlexiaJM I have the same bug. But installing the dev-build from github of 'trl' and the latest pip version of 'datasets' doesn't work for me.

**Comment by jayachandrakalakutagar** [comment]
> @AlexiaJM I have the same bug. But installing the dev-build from github of 'trl' and the latest pip version of 'datasets' doesn't work for me.

HI were you able to solve it? i am pretty sure we are making some mistake because on my first iteration i had got good results

**Comment by 0nutation** [comment]
> > @AlexiaJM I have the same bug. But installing the dev-build from github of 'trl' and the latest pip version of 'datasets' doesn't work for me.
> 
> HI were you able to solve it? i am pretty sure we are making some mistake because on my first iteration i had got good results

I don't solve it now. What do you mean by "my first iteration i had got good results" ? Why does this indicate some mistake? Where do you think the mistake might come from?

**Comment by 0nutation** [comment]
> > @AlexiaJM I have the same bug. But installing the dev-build from github of 'trl' and the latest pip version of 'datasets' doesn't work for me.
> 
> HI were you able to solve it? i am pretty sure we are making some mistake because on my first iteration i had got good results

Do you solve it?

**Comment by fc2869** [comment]
I had the same issue, getting the exact same loss as OP had (0.6932) for a couple of epochs. I resolved it by upgrading trl version from 0.7.4 to 0.7.11 and also lowered the learning rate.

**Comment by WeiXiongUST** [comment]
Hi, I was working on the reward modeling with the mistral in the past few weeks and encounter the same issue. The problem in my case is that the standard Chat-template prevent the models from handling multi-round chat if you set the pad token as the eon token. Instead, the following modification solves the problem.

tokenizer.add_special_tokens({'pad_token': '[PAD]'})
model.resize_token_embeddings(len(tokenizer))

Hope this helps!

**Comment by virt9** [comment]
hello！ i meet the same problem , my loss stuck at 0.6931 and the rewards is 0.0 ,i set ref_model is None and my learning rate is 1e-6 ,do you solve the problem?

**Comment by younesbelkada** [comment]
From the discussions above and internally, this could be solved by tweaking hyper-parameters. Can you try to play a bit with the learning rate and let us know how it goes?

**Comment by lixiaochuan2020** [comment]
@FangHainannn 's solution works. Whether we use create_reference_model() or copy.deepcopy(), or just leave the "ref_model" in DPOTrainer empty(if it is empty, huggingface will create a copy for you as the initial model), all work. 

The mistake for me is I wrongly passed the "model" to the "ref_model" parameter, causing the 'model' and 'ref_model' to point to the same object.

**Comment by kencyshaka** [comment]
@0nutation Did you solve this?
